stellargraph/layer/deep_graph_infomax.py

```diff
@@ -0,0 +1,137 @@
+
+    def build(self, input_shapes):
+
+        self.kernel = self.add_weight(
+            shape=(input_shapes[0][1], input_shapes[0][1]),
+            initializer="glorot_uniform",
+            name="kernel",
+            regularizer=None,
+            constraint=None,
+        )
+        self.built = True
+
+    def call(self, inputs):
+        """
+        Applies the layer to the inputs.
+
+        Args:
+            inputs: a list of tensors with shapes [(N, F), (F,)] containing the node features and summary feature
+                vector.
+        """
+        features, summary = inputs
+
+        score = tf.linalg.matvec(tf.linalg.matmul(features, self.kernel), summary)
+
+        return score
+
+
+@experimental(reason="lack of unit tests", issues=[1003])
+def fullbatch_infomax_node_model(base_model):
+    """
+    A function to create the the inputs and outputs for a Deep Graph Infomax model for unsupervised training.
+
+    Args:
+        base_model: the base stellargraph model class
+
+    Returns:
+        input and output layers for use with a keras model
+    """
+
+    if not isinstance(base_model, (GCN, GAT, APPNP, PPNP)):
+        raise TypeError(
+            f"base_model: expected GCN, GAT, APPNP or PPNP found {type(base_model).__name__}"
+        )
+
+    # Inputs for features
+    x_t = Input(batch_shape=(1, base_model.n_nodes, base_model.n_features))
+    # Inputs for shuffled features
+    x_corr = Input(batch_shape=(1, base_model.n_nodes, base_model.n_features))
+
+    out_indices_t = Input(batch_shape=(1, None), dtype="int32")
+
+    # Create inputs for sparse or dense matrices
+    if base_model.use_sparse:
+        # Placeholders for the sparse adjacency matrix
+        A_indices_t = Input(batch_shape=(1, None, 2), dtype="int64")
+        A_values_t = Input(batch_shape=(1, None))
+        A_placeholders = [A_indices_t, A_values_t]
+
+    else:
+        # Placeholders for the dense adjacency matrix
+        A_m = Input(batch_shape=(1, base_model.n_nodes, base_model.n_nodes))
+        A_placeholders = [A_m]
+
+    x_inp = [x_corr, x_t, out_indices_t] + A_placeholders
+
+    node_feats = base_model([x_t, out_indices_t] + A_placeholders)
+    node_feats = Lambda(lambda x: K.squeeze(x, axis=0), name=_NODE_FEATS,)(node_feats)
+
+    node_feats_corrupted = base_model([x_corr, out_indices_t] + A_placeholders)
+    node_feats_corrupted = Lambda(lambda x: K.squeeze(x, axis=0))(node_feats_corrupted)
+
+    summary = Lambda(lambda x: tf.math.sigmoid(tf.math.reduce_mean(x, axis=0)))(
+        node_feats
+    )
+
+    discriminator = Discriminator()
+    scores = discriminator([node_feats, summary])
+    scores_corrupted = discriminator([node_feats_corrupted, summary])
+
+    x_out = tf.stack([scores, scores_corrupted], axis=1)
+
+    x_out = K.expand_dims(x_out, axis=0)
+    return x_inp, x_out
+
+
+@experimental(reason="lack of unit tests", issues=[1003])
+def fullbatch_infomax_embedding_model(info_max_model):
```

Comment:
Why is this and `fullbatch_infomax_node_model` implemented as methods? If we are to be consistent with the rest of the library, we should probably have  a `DeepGraphInfomax` class that we can initialise as necessary to implement the different variations of the method in the paper such as for example passing the base model (GCN, GAT, GraphSAGE, etc) and the setting (single or multi-graph) just to name a couple of options.

Reply:
I replaced this with a class in a commit pushed just before your comments - but I think this mustve been after you started the review process!

Reply:
I just noticed that this has changed. The version I was reviewing became obsolete while I was reviewing it!

Reply:
I stopped pushing when I realised you were reviewing, but I think at the point the damage had been done!